crypto/threads_pthread.c

```diff
@@ -42,6 +42,404 @@
 #  define USE_RWLOCK
 # endif
 
+static pthread_key_t rcu_thr_key;
+
+struct rcu_thr_data;
+
+/*
+ * users is broken up into 2 parts
+ * bits 0-15 current readers
+ * bit 32-63 - ID
+ */
+# define READER_SHIFT 0
+# define ID_SHIFT 32
+# define READER_SIZE 16
+# define ID_SIZE 32
+
+# define READER_MASK     (((uint64_t)1 << READER_SIZE)-1)
+# define ID_MASK         (((uint64_t)1 << ID_SIZE)-1)
+# define READER_COUNT(x) (((uint64_t)(x) >> READER_SHIFT) & READER_MASK)
+# define ID_VAL(x)       (((uint64_t)(x) >> ID_SHIFT) & ID_MASK)
+# define VAL_READER      ((uint64_t)1 << READER_SHIFT)
+# define VAL_ID(x)       ((uint64_t)x << ID_SHIFT)
+
+/*
+ * This is the core of an rcu lock
+ * it tracks the the readers and writers
+ * for the current quiescence point for
+ * a given lock
+ * users is the 64 bit value that stores
+ * the READERS/ID as defined
+ * above
+ *
+ */
+struct rcu_qp {
+    uint64_t users;
+};
+
+/*
+ * This is the per thread tracking data
+ * that is assigned to each thread participating
+ * in an rcu qp
+ *
+ * qp points to the qp that it last acquired
+ *
+ * count is the threads current recursion count
+ */
+struct rcu_thr_data {
+    struct rcu_qp *qp;
+    int count;
+};
+
+/*
+ * This is the internal version of a CRYPTO_RCU_LOCK
+ * it is cast from CRYPTO_RCU_LOCK
+ */
+struct rcu_lock_internal {
+    CRYPTO_RCU_LOCK lock; /* Must be first */
+    struct rcu_cb_item *cb_items;
+    uint32_t id_ctr;
```

Comment:
Are these duplicates? They are defined in CRYPTO_RCU_LOCK.

Reply:
they are, and have since been de-duplicated (they belong in the internal structure)